### PR TITLE
Rewrite `non_exhaustive_omitted_patterns` as a separate pass

### PR DESCRIPTION
**Pull request by Nadrieril**
This is a rework of the `non_exhaustive_omitted_patterns` lint. It changes the semantics of the lint for two reasons:
- the previous implementation was inconsistent and hard to make consistent without impacting performance of exhaustiveness checking in general (discussed [here](https://github.com/rust-lang/rust/issues/89554#issuecomment-1548582793));
- this matches more closely the intent of the lint as I understand it, which is to help users keep their code up-to-date with upstream changes.

The new lint ignores considerations of exhaustiveness. Instead it looks at any part of the match that mentions some variants of a non_exhaustive enum, and triggers the lint if the match doesn't mention all the variants. That way users can ensure they update their code to mention all variants.
I also fixed some other inconsistencies in how the lint was implemented.

This separate pass is probably slow (but optional). I wouldn't recommend enabling the lint crate-wide. That said, correctness comes before performance, and I have plans that should make it much better.

### CODE REVIEW

**Comment by eholk** [comment]
I'm starting to look over this now, but let me first ask to make sure I understand the intent of the lint. The idea is that you want to make sure you're matching against all currently known variants of an enum, right? In other words, `#[non_exhaustive]` means you must have a wildcard, while this lint makes sure that you are still doing an exhaustive match, even if it wouldn't be a compilation error if more variants were added in the future. Is that the idea?

**Review by eholk — COMMENTED** [review]
Still working on this, but I need to take a break so I wanted to post the couple of comments I have so far.

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 1]
```diff
@@ -3814,9 +3816,9 @@ declare_lint! {
     /// // in crate B
     /// #![feature(non_exhaustive_omitted_patterns_lint)]
     ///
+    /// #[warn(non_exhaustive_omitted_patterns)]
     /// match Bar::A {
     ///     Bar::A => {},
-    ///     #[warn(non_exhaustive_omitted_patterns)]
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply compiler/rustc_lint_defs/src/builtin.rs:3819]: Not anymore, that made everything more complicated. The annotation must be on something that contains the whole match now

**Comment by Nadrieril** [comment]
Yeah I think you got the idea. It's basically a "please remind me if someone adds a variant to this enum because I want to handle all of them".

Very annoyed at the perf result, how can it be worse I just removed code and added a single top-level check :'( Will have to investigate

File: compiler/rustc_mir_build/src/thir/pattern/usefulness.rs [hunk 2]
```diff
@@ -642,14 +642,7 @@ impl<'p, 'tcx> Usefulness<'p, 'tcx> {
                         .into_iter()
                         .flat_map(|witness| {
                             new_patterns.iter().map(move |pat| {
-                                Witness(
-                                    witness
-                                        .0
-                                        .iter()
-                                        .chain(once(pat))
-                                        .map(DeconstructedPat::clone_and_forget_reachability)
```
Comment (eholk) [inline compiler/rustc_mir_build/src/thir/pattern/usefulness.rs:650]: Why is it okay not to forget reachability now? Is there a risk of having stale or incomplete reachability information?

File: tests/ui/rfc-2008-non-exhaustive/omitted-patterns.rs [hunk 3]
```diff
@@ -31,122 +32,115 @@ pub enum Bar {
...
 
     // Ok: because this only has 1 variant
-    #[deny(non_exhaustive_omitted_patterns)]
     match NonExhaustiveSingleVariant::A(true) {
         NonExhaustiveSingleVariant::A(true) => {}
         _ => {}
     }
 
-    #[deny(non_exhaustive_omitted_patterns)]
+    // Don't lint if no variant is mentioned, because we can't do it consistently.
```
Comment (eholk) [inline tests/ui/rfc-2008-non-exhaustive/omitted-patterns.rs:120]: Interesting! Why not?

**Review by eholk — APPROVED** [review]
This looks pretty good to me! I'll hold off on sending it to bors so you can chase the performance issues if you want.

I did have a couple questions that I added.

File: compiler/rustc_mir_build/src/thir/pattern/usefulness.rs (hunk 2 above)
Comment (Nadrieril) [reply compiler/rustc_mir_build/src/thir/pattern/usefulness.rs:650]: Nah this changes nothing, it's just that I hadn't implemented Clone to make sure no one cloned stuff by mistake. But it's inconvenient and there's not sensible situation where one would get confused. I have plans to clear that up more in the future

File: tests/ui/rfc-2008-non-exhaustive/omitted-patterns.rs (hunk 3 above)
Comment (Nadrieril) [reply tests/ui/rfc-2008-non-exhaustive/omitted-patterns.rs:120]: Take for example
```rust
match Some(&NonExhaustiveEnum::A) {
  _ => {}
}
```
In order to lint here I'd need to know that the enum exists at all, which means I'd need to dig into the `_` pattern, before even knowing if there might be an interesting enum somewhere deep. I can't just dig into all wildcards like that though, that would destroy performance.

Another approach would be to somehow compute ahead of time for all types whether they have some non_exhaustive enum deep down, but that doesn't seem worth the effort.

**Comment by Nadrieril** [comment]
Alright so I'm kinda rewriting the exhaustiveness algorithm big time over in https://github.com/rust-lang/rust/pull/111720 x). It's a rewrite I've been itching to do for years but could never get right. If it works this time, the performance argument I stated in the OP is moot. In fact what I'm doing here is a lot slower because it's a whole separate pass.

So hum I don't know what to do anymore. There's still a discussion to be had about what we want the lint to do but I don't think this PR is the right solution anymore. I'll close this for now and maybe come back to it when I'm done with the other one. Thanks for taking the time to review!

**Comment by Nadrieril** [comment]
I revived this in https://github.com/rust-lang/rust/pull/116734

**Pull request closed without merging**